### PR TITLE
feat(vuln): Add `--detection-priority` flag for accuracy tuning

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
This PR introduces a new `--detection-priority` flag to Trivy, allowing users to control the balance between false positives and false negatives in vulnerability detection. This feature enhances Trivy's flexibility in different use cases and risk tolerance levels.

#### `--detection-priority precise`

```
$ trivy image registry.access.redhat.com/ubi8/ubi --detection-priority precise --scanners vuln
```

<details><summary>Result</summary>

```
2024-08-01T11:16:55+04:00       INFO    [vuln] Vulnerability scanning is enabled
2024-08-01T11:17:02+04:00       INFO    Detected OS     family="redhat" version="8.10"
2024-08-01T11:17:02+04:00       INFO    [redhat] Detecting RHEL/CentOS vulnerabilities...       os_version="8" pkg_num=183
2024-08-01T11:17:02+04:00       INFO    Number of language-specific files       num=0

registry.access.redhat.com/ubi8/ubi (redhat 8.10)
=================================================
Total: 180 (UNKNOWN: 0, LOW: 160, MEDIUM: 20, HIGH: 0, CRITICAL: 0)
```
</details> 

#### `--detection-priority comprehensive` 

```
$ trivy image registry.access.redhat.com/ubi8/ubi --detection-priority comprehensive --scanners vuln
```

<details><summary>Result</summary>

```
2024-08-01T11:18:16+04:00       INFO    [vuln] Vulnerability scanning is enabled
2024-08-01T11:18:23+04:00       INFO    Detected OS     family="redhat" version="8.10"
2024-08-01T11:18:23+04:00       INFO    [redhat] Detecting RHEL/CentOS vulnerabilities...       os_version="8" pkg_num=183
2024-08-01T11:18:23+04:00       INFO    Number of language-specific files       num=1
2024-08-01T11:18:23+04:00       INFO    [python-pkg] Detecting vulnerabilities...

registry.access.redhat.com/ubi8/ubi (redhat 8.10)
=================================================
Total: 180 (UNKNOWN: 0, LOW: 160, MEDIUM: 20, HIGH: 0, CRITICAL: 0)

...

2024-08-01T11:18:23+04:00       INFO    Table result includes only package filenames. Use '--format json' option to get the full path to the package file.

Python (python-pkg)
===================
Total: 10 (UNKNOWN: 0, LOW: 0, MEDIUM: 8, HIGH: 2, CRITICAL: 0)

┌───────────────────────┬────────────────┬──────────┬────────┬───────────────────┬────────────────┬─────────────────────────────────────────────────────────────┐
│        Library        │ Vulnerability  │ Severity │ Status │ Installed Version │ Fixed Version  │                            Title                            │
├───────────────────────┼────────────────┼──────────┼────────┼───────────────────┼────────────────┼─────────────────────────────────────────────────────────────┤
│ idna (PKG-INFO)       │ CVE-2024-3651  │ MEDIUM   │ fixed  │ 2.5               │ 3.7            │ python-idna: potential DoS via resource consumption via     │
│                       │                │          │        │                   │                │ specially crafted inputs to idna.encode()...                │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2024-3651                   │
├───────────────────────┼────────────────┤          │        ├───────────────────┼────────────────┼─────────────────────────────────────────────────────────────┤
│ requests (PKG-INFO)   │ CVE-2023-32681 │          │        │ 2.20.0            │ 2.31.0         │ python-requests: Unintended leak of Proxy-Authorization     │
│                       │                │          │        │                   │                │ header                                                      │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2023-32681                  │
│                       ├────────────────┤          │        │                   ├────────────────┼─────────────────────────────────────────────────────────────┤
│                       │ CVE-2024-35195 │          │        │                   │ 2.32.0         │ requests: subsequent requests to the same host ignore cert  │
│                       │                │          │        │                   │                │ verification                                                │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2024-35195                  │
├───────────────────────┼────────────────┼──────────┤        ├───────────────────┼────────────────┼─────────────────────────────────────────────────────────────┤
│ setuptools (METADATA) │ CVE-2022-40897 │ HIGH     │        │ 39.2.0            │ 65.5.1         │ pypa-setuptools: Regular Expression Denial of Service       │
│                       │                │          │        │                   │                │ (ReDoS) in package_index.py                                 │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2022-40897                  │
│                       ├────────────────┤          │        │                   ├────────────────┼─────────────────────────────────────────────────────────────┤
│                       │ CVE-2024-6345  │          │        │                   │ 70.0.0         │ pypa/setuptools: Remote code execution via download         │
│                       │                │          │        │                   │                │ functions in the package_index module in...                 │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2024-6345                   │
├───────────────────────┼────────────────┼──────────┤        ├───────────────────┼────────────────┼─────────────────────────────────────────────────────────────┤
│ urllib3 (PKG-INFO)    │ CVE-2019-11236 │ MEDIUM   │        │ 1.24.2            │ 1.24.3         │ python-urllib3: CRLF injection due to not encoding the      │
│                       │                │          │        │                   │                │ '\r\n' sequence leading to...                               │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2019-11236                  │
│                       ├────────────────┤          │        │                   ├────────────────┼─────────────────────────────────────────────────────────────┤
│                       │ CVE-2020-26137 │          │        │                   │ 1.25.9         │ python-urllib3: CRLF injection via HTTP request method      │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2020-26137                  │
│                       ├────────────────┤          │        │                   ├────────────────┼─────────────────────────────────────────────────────────────┤
│                       │ CVE-2023-43804 │          │        │                   │ 2.0.6, 1.26.17 │ python-urllib3: Cookie request header isn't stripped during │
│                       │                │          │        │                   │                │ cross-origin redirects                                      │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2023-43804                  │
│                       ├────────────────┤          │        │                   ├────────────────┼─────────────────────────────────────────────────────────────┤
│                       │ CVE-2023-45803 │          │        │                   │ 2.0.7, 1.26.18 │ urllib3: Request body not stripped after redirect from 303  │
│                       │                │          │        │                   │                │ status changes request...                                   │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2023-45803                  │
│                       ├────────────────┤          │        │                   ├────────────────┼─────────────────────────────────────────────────────────────┤
│                       │ CVE-2024-37891 │          │        │                   │ 1.26.19, 2.2.2 │ urllib3: proxy-authorization request header is not stripped │
│                       │                │          │        │                   │                │ during cross-origin redirects                               │
│                       │                │          │        │                   │                │ https://avd.aquasec.com/nvd/cve-2024-37891                  │
└───────────────────────┴────────────────┴──────────┴────────┴───────────────────┴────────────────┴─────────────────────────────────────────────────────────────┘
```

</details> 

#### `--detection-priority precise`

```
$ trivy image cgr.dev/chainguard/argocd --detection-priority precise --scanners vuln
```

<details><summary>Result</summary>

```
2024-08-01T11:27:12+04:00       INFO    [vuln] Vulnerability scanning is enabled
2024-08-01T11:27:12+04:00       INFO    Detected OS     family="wolfi" version="20230201"
2024-08-01T11:27:12+04:00       INFO    [wolfi] Detecting vulnerabilities...    pkg_num=10
2024-08-01T11:27:12+04:00       INFO    Number of language-specific files       num=0
2024-08-01T11:27:12+04:00       WARN    Using severities from other vendors for some vulnerabilities. Read https://aquasecurity.github.io/trivy/dev/docs/scanner/vulnerability#severity-selection for details.

cgr.dev/chainguard/argocd (wolfi 20230201)
==========================================
Total: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 2, CRITICAL: 0)

┌─────────────────────┬────────────────┬──────────┬────────┬───────────────────┬───────────────┬────────────────────────────────────────────────────┐
│       Library       │ Vulnerability  │ Severity │ Status │ Installed Version │ Fixed Version │                       Title                        │
├─────────────────────┼────────────────┼──────────┼────────┼───────────────────┼───────────────┼────────────────────────────────────────────────────┤
│ argo-cd-2.11        │ CVE-2024-40634 │ HIGH     │ fixed  │ 2.11.5-r0         │ 2.11.6-r0     │ argocd: Unauthenticated Denial of Service (DoS)    │
│                     │                │          │        │                   │               │ Vulnerability via /api/webhook Endpoint in Argo... │
│                     │                │          │        │                   │               │ https://avd.aquasec.com/nvd/cve-2024-40634         │
├─────────────────────┤                │          │        │                   │               │                                                    │
│ argo-cd-2.11-compat │                │          │        │                   │               │                                                    │
│                     │                │          │        │                   │               │                                                    │
│                     │                │          │        │                   │               │                                                    │
└─────────────────────┴────────────────┴──────────┴────────┴───────────────────┴───────────────┴────────────────────────────────────────────────────┘
```

</details>

#### `--detection-priority comprehensive`

```
$ trivy image cgr.dev/chainguard/argocd --detection-priority comprehensive --scanners vuln
```

<details><summary>Result</summary>

```
2024-08-01T11:27:54+04:00       INFO    [vuln] Vulnerability scanning is enabled
2024-08-01T11:27:54+04:00       INFO    Detected OS     family="wolfi" version="20230201"
2024-08-01T11:27:54+04:00       INFO    [wolfi] Detecting vulnerabilities...    pkg_num=10
2024-08-01T11:27:54+04:00       INFO    Number of language-specific files       num=2
2024-08-01T11:27:54+04:00       INFO    [gobinary] Detecting vulnerabilities...
2024-08-01T11:27:54+04:00       WARN    Using severities from other vendors for some vulnerabilities. Read https://aquasecurity.github.io/trivy/dev/docs/scanner/vulnerability#severity-selection for details.

cgr.dev/chainguard/argocd (wolfi 20230201)
==========================================
Total: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 2, CRITICAL: 0)

┌─────────────────────┬────────────────┬──────────┬────────┬───────────────────┬───────────────┬────────────────────────────────────────────────────┐
│       Library       │ Vulnerability  │ Severity │ Status │ Installed Version │ Fixed Version │                       Title                        │
├─────────────────────┼────────────────┼──────────┼────────┼───────────────────┼───────────────┼────────────────────────────────────────────────────┤
│ argo-cd-2.11        │ CVE-2024-40634 │ HIGH     │ fixed  │ 2.11.5-r0         │ 2.11.6-r0     │ argocd: Unauthenticated Denial of Service (DoS)    │
│                     │                │          │        │                   │               │ Vulnerability via /api/webhook Endpoint in Argo... │
│                     │                │          │        │                   │               │ https://avd.aquasec.com/nvd/cve-2024-40634         │
├─────────────────────┤                │          │        │                   │               │                                                    │
│ argo-cd-2.11-compat │                │          │        │                   │               │                                                    │
│                     │                │          │        │                   │               │                                                    │
│                     │                │          │        │                   │               │                                                    │
└─────────────────────┴────────────────┴──────────┴────────┴───────────────────┴───────────────┴────────────────────────────────────────────────────┘

usr/bin/argocd (gobinary)
=========================
Total: 4 (UNKNOWN: 0, LOW: 1, MEDIUM: 2, HIGH: 1, CRITICAL: 0)

┌──────────────────────────────────────────────────┬────────────────┬──────────┬────────┬───────────────────┬──────────────────────────────────┬───────────────────────────────────────────────────────────┐
│                     Library                      │ Vulnerability  │ Severity │ Status │ Installed Version │          Fixed Version           │                           Title                           │
├──────────────────────────────────────────────────┼────────────────┼──────────┼────────┼───────────────────┼──────────────────────────────────┼───────────────────────────────────────────────────────────┤
│ github.com/Azure/azure-sdk-for-go/sdk/azidentity │ CVE-2024-35255 │ MEDIUM   │ fixed  │ v1.1.0            │ 1.6.0                            │ azure-identity: Azure Identity Libraries Elevation of     │
│                                                  │                │          │        │                   │                                  │ Privilege Vulnerability in                                │
│                                                  │                │          │        │                   │                                  │ github.com/Azure/azure-sdk-for-go/sdk/azidentity          │
│                                                  │                │          │        │                   │                                  │ https://avd.aquasec.com/nvd/cve-2024-35255                │
├──────────────────────────────────────────────────┼────────────────┼──────────┤        ├───────────────────┼──────────────────────────────────┼───────────────────────────────────────────────────────────┤
│ github.com/argoproj/argo-cd/v2                   │ CVE-2024-40634 │ HIGH     │        │ 2.11.5            │ 2.9.20, 2.10.15, 2.11.6          │ argocd: Unauthenticated Denial of Service (DoS)           │
│                                                  │                │          │        │                   │                                  │ Vulnerability via /api/webhook Endpoint in Argo...        │
│                                                  │                │          │        │                   │                                  │ https://avd.aquasec.com/nvd/cve-2024-40634                │
├──────────────────────────────────────────────────┼────────────────┼──────────┤        ├───────────────────┼──────────────────────────────────┼───────────────────────────────────────────────────────────┤
│ k8s.io/kubernetes                                │ CVE-2024-5321  │ MEDIUM   │        │ v1.26.11          │ 1.27.16, 1.28.12, 1.29.7, 1.30.3 │ kubelet: Incorrect permissions on Windows containers logs │
│                                                  │                │          │        │                   │                                  │ https://avd.aquasec.com/nvd/cve-2024-5321                 │
│                                                  ├────────────────┼──────────┤        │                   ├──────────────────────────────────┼───────────────────────────────────────────────────────────┤
│                                                  │ CVE-2024-3177  │ LOW      │        │                   │ 1.27.13, 1.29.4, 1.28.9          │ kubernetes: kube-apiserver: bypassing mountable secrets   │
│                                                  │                │          │        │                   │                                  │ policy imposed by the ServiceAccount admission plugin...  │
│                                                  │                │          │        │                   │                                  │ https://avd.aquasec.com/nvd/cve-2024-3177                 │
└──────────────────────────────────────────────────┴────────────────┴──────────┴────────┴───────────────────┴──────────────────────────────────┴───────────────────────────────────────────────────────────┘

...
```

</details> 

### Notes
1. OS packages:
   - When OS vendors provide sufficient advisories (e.g., Red Hat), `--detection-priority comprehensive` may not significantly increase true positive detections. Newly detected vulnerabilities are likely to be false positives or duplicates of vulnerabilities already detected via OS vendor advisories (see the above example).
   - For cases where OS vendor advisories are insufficient or delayed compared to upstream advisories, the comprehensive mode may detect more relevant vulnerabilities.

2. Language-specific packages:
   - In cases where exact versions cannot be determined (e.g., version ranges), `--detection-priority comprehensive` uses the lower bound of version ranges for vulnerability detection. This detects possible vulnerabilities, but may lead to false positives if the actual version used is not at this lower bound.

### Key changes
- Implemented the `--detection-priority` flag with two modes: `precise` and `comprehensive`
- Added logic to adjust vulnerability detection based on the chosen mode
- Updated relevant tests to cover the new functionality
- Added documentation explaining the new flag and its usage

### Implementation details
- `precise` mode (default): Focuses on reducing false positives, resulting in less noisy vulnerability reports
- `comprehensive` mode: Aims to detect more vulnerabilities, potentially including some false positives, for broader coverage
    - Analyze files installed by an OS package manager
    - Use a minimum version of the version ranges
- Included brief user review considerations for each mode

## Terminology
I chose `comprehensive` instead of `coverage` for the following reasons:

1. Clarity of meaning: `comprehensive` better describes the broader detection approach
1. Avoiding confusion with the existing [Scanning Coverage](https://aquasecurity.github.io/trivy/v0.54/docs/coverage/) documentation
1. Distinction from technical terms: `coverage` has specific meanings in testing contexts

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/7270

## Related PRs
- https://github.com/aquasecurity/trivy/pull/4109

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: docs/docs/scanner/vulnerability.md [hunk 1]
```diff
@@ -307,6 +307,25 @@ By default, all relationships are included in the scan.
...
+- `precise`: This mode prioritizes reducing false positives. It results in less noisy vulnerability reports but may miss some potential vulnerabilities.
+- `comprehensive`: This mode aims to detect more vulnerabilities, potentially including some that might be false positives.
+  It provides broader coverage but may increase the noise in the results.
+
+The default value is `precise`.
+
+Regardless of the chosen mode, user review of detected vulnerabilities is crucial:
+
+- `precise`: Review thoroughly, considering potential missed vulnerabilities.
+- `comprehensive`: Carefully investigate each reported vulnerability due to increased false positive possibility.
```
Comment (DmitriyLewen) [inline docs/docs/scanner/vulnerability.md:328]: I think we need to add `Notes` from PR description to docs.
We can also add a table with differences.
something like this:
|   | precise   | comprehensive  | 
|---|---|---|
| pkg managers  | only OS pkgs | OS pkgs + language pkgs |
| dart  | `0.0.0`  | use min version from SDK |
| gomod  | skip `stdlib`  | use min version from `toolchain` |

File: pkg/cache/key_test.go
Comment (DmitriyLewen) [inline pkg/cache/key_test.go:1]: I suggest adding test with non-empty `DetectionPriority` value.

**Review by DmitriyLewen — COMMENTED** [review]
LGTM
Left a couple of comments.

File: docs/docs/scanner/vulnerability.md (hunk 1 above)
Comment (knqyf263) [reply docs/docs/scanner/vulnerability.md:376]: Updated https://github.com/aquasecurity/trivy/pull/7288/commits/67b80d32b9adf21c8474fa8622ada35ace98ab9c

File: pkg/cache/key_test.go
Comment (knqyf263) [reply pkg/cache/key_test.go:1]: Added https://github.com/aquasecurity/trivy/pull/7288/commits/1ce2e4bad6f2504afaec02c7d65954a6d0297adf

File: docs/docs/coverage/language/python.md [hunk 2]
```diff
@@ -21,11 +21,11 @@ The following scanners are supported for Python packages.
...
-|-----------------|------------------|:-----------------------:|:----------------:|:------------------------------------:|:--------:|
-| pip             | requirements.txt |            -            |     Include      |                  -                   |    ✓     |
-| Pipenv          | Pipfile.lock     |            ✓            |     Include      |                  -                   |    ✓     |
-| Poetry          | poetry.lock      |            ✓            |     Exclude      |                  ✓                   |    -     |
+| Package manager | File             | Transitive dependencies | Dev dependencies | [Dependency graph][dependency-graph] | Position | [Detection Priority][detection-priority] |
+|-----------------|------------------|:-----------------------:|:----------------:|:------------------------------------:|:--------:|:----------------------------------------:|
+| pip             | requirements.txt |            -            |     Include      |                  -                   |    ✓     |                    -                     |
```
Comment (DmitriyLewen) [inline docs/docs/coverage/language/python.md:26]: we can add table for conda `environment.yml` files (this file list `requirements.txt`)

File: docs/docs/coverage/language/java.md [hunk 3]
```diff
@@ -12,12 +12,12 @@ Each artifact supports the following scanners:
...
-| JAR/WAR/PAR/EAR  |     Trivy Java DB     |     Include      |                  -                   |    -     |
-| pom.xml          | Maven repository [^1] |     Exclude      |                  ✓                   |  ✓[^7]   |
-| *gradle.lockfile |           -           |     Exclude      |                  ✓                   |    ✓     |
-| *.sbt.lock       |          -            |     Exclude      |                  -                   |    ✓     |
+| Artifact         |    Internet access    | Dev dependencies | [Dependency graph][dependency-graph] | Position | [Detection Priority][detection-priority] |
+|------------------|:---------------------:|:----------------:|:------------------------------------:|:--------:|:----------------------------------------:|
+| JAR/WAR/PAR/EAR  |     Trivy Java DB     |     Include      |                  -                   |    -     |                Not needed                |
+| pom.xml          | Maven repository [^1] |     Exclude      |                  ✓                   |  ✓[^7]   |                    -                     |
```
Comment (DmitriyLewen) [inline docs/docs/coverage/language/java.md:18]: Can you remind me where in `pom` files you want to use new flag?

Comment (knqyf263) [reply docs/docs/coverage/language/java.md:18]: We take soft and hard requirements now, but Maven has more requirements, such as `[1.0,)` and `[1.0, 2)`.
https://github.com/aquasecurity/trivy/blob/e95152f796308c25aaaa6cf75b2e7a81b3ab4388/pkg/dependency/parser/java/pom/artifact.go#L89

Comment (DmitriyLewen) [reply docs/docs/coverage/language/java.md:18]: Exactly! Thanks for reminding me

File: docs/docs/coverage/language/python.md (hunk 2 above)
Comment (knqyf263) [reply docs/docs/coverage/language/python.md:26]: Did we forget to add `environment.yaml` to the doc?

Comment (knqyf263) [reply docs/docs/coverage/language/python.md:26]: Oh, I found it under OS. Do you remember why we put it here?
https://aquasecurity.github.io/trivy/v0.54/docs/coverage/os/conda/

Comment (DmitriyLewen) [reply docs/docs/coverage/language/python.md:26]: `Conda` doesn't only contain `Python` packages.
You can install OS packages.
`Conda` is thus compilation of package manager (like `apt`) and `pip`.

I thought about adding `Conda` to package managers, but we don't have pages for that.
So we decided to put `Conda` in OS first.

Comment (knqyf263) [reply docs/docs/coverage/language/python.md:26]: >You can install OS packages.

Is there any document about it?

Anyway, I added.
https://github.com/aquasecurity/trivy/pull/7288/commits/9e8cdb17910410bcf82cc726ef5d577c1e0c1862

Comment (DmitriyLewen) [reply docs/docs/coverage/language/python.md:26]: https://docs.anaconda.com/working-with-conda/packages/install-packages/#installing-conda-packages

e.g. you can install `curl` - https://anaconda.org/conda-forge/curl

**Review by DmitriyLewen — APPROVED** [review]
LGTM

File: docs/docs/coverage/language/python.md (hunk 2 above)
Comment (knqyf263) [reply docs/docs/coverage/language/python.md:26]: I have reconsidered this, but my personal definition of "OS packages" is a package distributed by an OS vendor, such as Red Hat and SUSE. Conda is simply building and distributing third-party software, which is different from an operating system. Bitnami does the same and it seems better to create a new category in addition to OS and Language. We should discuss it later.

Comment (DmitriyLewen) [reply docs/docs/coverage/language/python.md:26]: We just didn't have a suitable category.
But in general we can separate `conda` + `bitnami` into a separate category.
